src/sentry/integrations/gitlab/webhooks.py

```diff
@@ -79,92 +142,79 @@ def create_commits(self, event, organization, repo):
     }
 
-    def get_handler(self, event_type):
-        return self._handlers.get(event_type)
-
     @method_decorator(csrf_exempt)
     def dispatch(self, request, *args, **kwargs):
         if request.method != 'POST':
             return HttpResponse(status=405)
 
-        # if not self.check_secret(request):
-        #     raise Exception???
-
         return super(GitlabWebhookEndpoint, self).dispatch(request, *args, **kwargs)
 
-    def post(self, request, organization_id):
+    def post(self, request):
         try:
-            organization = Organization.objects.get_from_cache(
-                id=organization_id,
-            )
-        except Organization.DoesNotExist:
-            logger.error(
-                PROVIDER_NAME + '.webhook.invalid-organization',
+            # Munge the token to extract the integration external_id.
+            # gitlab hook payloads don't give us enough unique context
+            # to find data on our side so we embed one in the token.
+            token = request.META['HTTP_X_GITLAB_TOKEN']
+            instance, group_path, secret = token.split(':')
+            external_id = u'{}:{}'.format(instance, group_path)
+        except Exception:
+            logger.info('gitlab.webhook.invalid-token', extra={'token': token})
+            return HttpResponse(status=400)
+
+        try:
+            integration = Integration.objects.filter(
+                provider=self.provider,
+                external_id=external_id
+            ).prefetch_related('organizations').get()
+        except Integration.DoesNotExist:
+            logger.info(
+                'gitlab.webhook.invalid-organization',
                 extra={
-                    'organization_id': organization_id,
+                    'external_id': request.META['HTTP_X_GITLAB_TOKEN'],
                 }
             )
             return HttpResponse(status=400)
 
-        body = six.binary_type(request.body)
-        if not body:
-            logger.error(
-                PROVIDER_NAME + '.webhook.missing-body', extra={
-                    'organization_id': organization.id,
+        if not constant_time_compare(secret, integration.metadata['webhook_secret']):
+            logger.info(
+                'gitlab.webhook.invalid-token-secret',
+                extra={
+                    'integration_id': integration.id
                 }
             )
             return HttpResponse(status=400)
 
         try:
-            handler = self.get_handler(request.META['HTTP_X_EVENT_KEY'])
-        except KeyError:
-            logger.error(
-                PROVIDER_NAME + '.webhook.missing-event', extra={
-                    'organization_id': organization.id,
+            event = json.loads(request.body.decode('utf-8'))
+        except JSONDecodeError:
+            logger.info(
+                'gitlab.webhook.invalid-json',
+                extra={
+                    'external_id': integration.external_id
```

Comment:
I'm not sure if this is too much, but would you add the request body here?

Reply:
The payload could be pretty big and I would be concerned with clogging up the pipe to Elasticsearch. Having the external_id should give enough information to get close to the problematic hook in GitLab.